### ISSUE TITLE
Overridden metaclass __call__ not taken into account for validity of parameters

### ISSUE THREAD

**Issue by alligatortower** [microsoft/pyright#3231]
**Describe the bug**
My metaclass implements a `__call__` method which takes an optional kwarg param and calls the obj `__new__` with that param, but not the `__init__`. pyright does not seem to be aware of this param.

**To Reproduce**
```python
class Meta(type):
    def __call__(cls, *args, param=None, **kwargs):
        obj = cls.__new__(*args, param=param, **kwargs)
        if obj is not None and isinstance(obj, cls) and hasattr(obj, '__init__'):
            obj.__init__(*args, **kwargs)
        return obj


class Parent(metaclass=Meta):
    def __new__(cls, param=None):
        if param:
            return super().__new__(SomeOtherCls)
        return super().__new__(cls)


class SomeOtherCls(metaclass=Meta):
    pass


class MyClass(Parent):
    def __init__(self):
        pass


instance = MyClass(param=True)   #  No parameter named "param"
```

**Expected behavior**
I expect `param` to be valid in the code above

**VS Code extension or command-line**
I'm running this via lunarvim / neovim lsp

**Comment by erictraut** [comment]
You're implementing non-standard behavior. Pyright (and other type checkers) need to make assumptions about certain behaviors. If you're going to deviate significantly from the normal behavior, you'll break these assumptions. If you cannot change your code to conform to more standard behaviors, you'll need to suppress the errors in some manner.

**Comment by Rapptz** [comment]
Should be noted the standard library does this with [enums](https://docs.python.org/3/library/enum.html#programmatic-access-to-enumeration-members-and-their-attributes).
